src/plugins/hidsim/gui_monitor.cpp

```diff
@@ -0,0 +1,443 @@
+ *     6.  Find the address to trap by subtracting the RVA of W32pServiceTable
+ *         from the VA of the GUI SSDT and then adding the function offset
+ *
+ * For a thorough description of the inner workings of the SSDT and
+ * the W32pServiceTable (aka GUI SSDT), see ERNWs blogpost at
+ * https://insinuator.net/2015/12/\
+ * investigating-memory-analysis-tools-ssdt-hooking-via-pointer-replacement/
+ */
+static bool register_trap( drakvuf_t drakvuf, json_object* profile_json,
+    const char* function_name, drakvuf_trap_t* trap,
+    event_response_t(*hook_cb)( drakvuf_t drakvuf, drakvuf_trap_info_t* info))
+{
+    addr_t func_rva = 0;
+    if (!json_get_symbol_rva(drakvuf, profile_json, function_name, &func_rva))
+    {
+        PRINT_DEBUG("[HIDSIM] [MONITOR] Failed to get RVA of win32k!%s\n",
+            function_name);
+        return false;
+    }
+
+    addr_t w32pst_rva = 0;
+    if (!json_get_symbol_rva(drakvuf, profile_json, "W32pServiceTable", &w32pst_rva))
+    {
+        PRINT_DEBUG("[HIDSIM] [MONITOR] Failed to get RVA of win32k!W32pServiceTable\n");
+        return false;
+    }
+
+    addr_t sdt_rva = 0;
+    if (!drakvuf_get_kernel_symbol_rva(drakvuf, "KeServiceDescriptorTableShadow", &sdt_rva))
+    {
+        PRINT_DEBUG("[HIDSIM] [MONITOR] Failed to get RVA of nt!KeServiceDescriptorTableShadow\n");
+        return false;
+    }
+
+    addr_t sdt_va = 0;
+    if (!(sdt_va = drakvuf_exportksym_to_va(drakvuf, 4, nullptr, "ntoskrnl.exe", sdt_rva)))
+    {
+        PRINT_DEBUG("[HIDSIM] [Init] Failed to get VA of nt!KeServiceDescriptorTableShadow\n");
+        return false;
+    }
+
+    const int SYSTEM_SERVICE_TABLE_32 = 16;
+    const int SYSTEM_SERVICE_TABLE_64 = 32;
+
+    bool is32bit = (drakvuf_get_page_mode(drakvuf) != VMI_PM_IA32E);
+    addr_t offset = is32bit ? SYSTEM_SERVICE_TABLE_32 : SYSTEM_SERVICE_TABLE_64;
+    addr_t ssdt_ptr_va = sdt_va + offset;
+
+    addr_t eprocess_base = 0;
+    if (!drakvuf_find_process(drakvuf, ~0, "explorer.exe", &eprocess_base))
+    {
+        PRINT_DEBUG("[HIDSIM] [MONITOR] Failed to find EPROCESS of \"y.exe\"\n");
+        return false;
+    }
+
+    vmi_pid_t pid = 0;
+    if (!drakvuf_get_process_pid(drakvuf, eprocess_base, &pid))
+    {
+        PRINT_DEBUG("[HIDSIM] [MONITOR] Failed to get PID of \"explorer.exe\"\n");
+        return false;
+    }
+
+    vmi_lock_guard vmi(drakvuf);
+
+    if (VMI_SUCCESS != vmi_pid_to_dtb(vmi.vmi, pid, &trap->breakpoint.dtb))
+    {
+        PRINT_DEBUG("[HIDSIM] [Init] Failed to get CR3 of \"explorer.exe\"\n");
+        return false;
+    }
+
+    ACCESS_CONTEXT(ctx,
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .addr = ssdt_ptr_va,
+        .dtb = trap->breakpoint.dtb
+    );
+
+    addr_t ssdt_va = 0;
+    if (VMI_SUCCESS != vmi_read_addr(vmi.vmi, &ctx, &ssdt_va))
+    {
+        PRINT_DEBUG("[HIDSIM] [MONITOR] Failed to read the address of SSDT (VA 0x%lx)\n", ssdt_ptr_va);
+        return false;
+    }
```

Comment:
Please add a scope here, no need to hold lock any longer than this.

Reply:
Added a scope.